x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/aggregation/AggregationToJsonProcessor.java

```diff
@@ -375,6 +455,37 @@ boolean writeDocs(int batchSize, OutputStream outputStream) throws IOException {
         return docsByBucketTimestamp.isEmpty() == false;
     }
 
+    /**
+     * This writes ALL the documents stored within the processor object unless indicated otherwise by the `shouldCancel` predicate
+     *
+     * This returns `true` if it is safe to cancel the overall process as the current `date_histogram` bucket has finished.
+     * @param shouldCancel determines if a given timestamp indicates that the processing stream should be cancelled
+     * @param outputStream where to write the aggregated data
+     * @return true if it is acceptable for the caller to close the process and cancel the stream
+     * @throws IOException if there is a parsing exception
+     */
+    boolean writeAllDocsCancellable(Predicate<Long> shouldCancel, OutputStream outputStream) throws IOException {
+        if (docsByBucketTimestamp.isEmpty()) {
+            return true;
+        }
+
+        try (XContentBuilder jsonBuilder = new XContentBuilder(JsonXContent.jsonXContent, outputStream)) {
+            Iterator<Map.Entry<Long, List<Map<String, Object>>>> iterator = docsByBucketTimestamp.entrySet().iterator();
+            while (iterator.hasNext()) {
+                Map.Entry<Long, List<Map<String, Object>>> entry = iterator.next();
+                if (shouldCancel.test(entry.getKey())) {
```

Comment:
OK, I thought about this and the discrepancy with the other version of `writeDocs`. I believe we can just keep the new version and use it where we used the old version as well.

Here is a bit of the story here.

The first implementation of allowing datafeeds with aggs to cancel added the idea of looking at how many key-value pairs we wrote. At that point, the implementation was not checking if we wrote the entire bucket. Thus, it was possible to stop the datafeed half through the bucket. This was fixed later, when the current implementation was added.

The current implementation collects data in buckets. We then write whole buckets and only check to see if we should cancel after a whole bucket was written. We still look for whether we reached the key-value pair batch size of `1000`. But I think this is completely unnecessary now.

The point of checking key-value pairs was to create a good moment to check whether we should cancel without having to check after each record. But checking after each histogram bucket is good enough. I hope this makes sense.

Reply:
Having said that, the other benefit of splitting is to contain the size of the output stream. But I guess this might be unnecessary as it is bound by the size of the search response, right? If the search response already fits in memory, then the output stream we create shouldn't be a problem.